reconcile/cna/assets/asset.py

```
@@ -1,38 +1,222 @@
     @abstractmethod
-    def api_payload(self) -> dict[str, Any]:
-        raise NotImplementedError()
+    def asset_type() -> AssetType:
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def provider() -> str:
+        ...
 
+    @staticmethod
     @abstractmethod
-    def update_from(self, asset: Asset) -> Asset:
-        raise NotImplementedError()
+    def from_query_class(asset: CNAssetV1) -> Asset:
+        ...
+
+    @staticmethod
+    def asset_type_from_raw_asset(raw_asset: Mapping[str, Any]) -> Optional[AssetType]:
+        asset_type_value = raw_asset[ASSET_TYPE_FIELD]
+        return asset_type_by_id(asset_type_value)
+
+    def asset_metadata(self) -> dict[str, Any]:
+        return {
+            "id": self.id,
+            "href": self.href,
+            "status": self.status.value if self.status else None,
+            "name": self.name,
+            ASSET_TYPE_FIELD: self.asset_type().value,
+        }
+
+    def api_payload(self) -> dict[str, Any]:
+        return {
+            ASSET_TYPE_FIELD: self.asset_type().value,
+            "name": self.name,
+            ASSET_PARAMETERS_FIELD: self.raw_asset_parameters(omit_empty=False),
+        }
+
+    def raw_asset_parameters(self, omit_empty: bool) -> dict[str, Any]:
+        raw_asset_params = {}
+        for var in self.type_metadata().variables:
+            python_property_name = _property_for_asset_parameter_alias(
+                type(self), var.name
+            )
+            var_value = getattr(self, python_property_name)
+            if not var.optional and var_value is None:
+                raise AssetError(
+                    f"Required variable {var.name} not set for asset {self.name}"
+                )
+            if var_value is not None or not omit_empty:
+                raw_asset_params[var.name] = var_value
+        return raw_asset_params
+
+    def update_from(
+        self,
+        asset: Asset,
+    ) -> Asset:
+        assert isinstance(asset, type(self))
```

should we raise an exception instead?